cmd/snapd-apparmor/main.go

```diff
@@ -0,0 +1,186 @@
+	if output, err := exec.Command("systemd-detect-virt", "--container").Output(); err == nil {
+		virt := strings.TrimSpace(string(output))
+		return virt == "wsl"
+	}
+	return false
+}
+
+// Checks to see if the current container is capable of having internal AppArmor
+// profiles that should be loaded.
+//
+// The only known container environments capable of supporting internal policy
+// are LXD and LXC environment.
+//
+// Returns true if the container environment is capable of having its own internal
+// policy and false otherwise.
+//
+// IMPORTANT: This function will return true in the case of a non-LXD/non-LXC
+// system container technology being nested inside of a LXD/LXC container that
+// utilized an AppArmor namespace and profile stacking. The reason true will be
+// returned is because .ns_stacked will be "yes" and .ns_name will still match
+// "lx[dc]-*" since the nested system container technology will not have set up
+// a new AppArmor profile namespace. This will result in the nested system
+// container's boot process to experience failed policy loads but the boot
+// process should continue without any loss of functionality. This is an
+// unsupported configuration that cannot be properly handled by this function.
+//
+func isContainerWithInternalPolicy() bool {
+	var appArmorSecurityFSPath = filepath.Join(dirs.GlobalRootDir, "/sys/kernel/security/apparmor")
+	var nsStackedPath = filepath.Join(appArmorSecurityFSPath, ".ns_stacked")
+	var nsNamePath = filepath.Join(appArmorSecurityFSPath, ".ns_name")
+
+	if isWSL() {
+		return true
+	}
+
+	for _, path := range []string{nsStackedPath, nsNamePath} {
+		if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
+			return false
+		}
+	}
+
+	contents, err := ioutil.ReadFile(nsStackedPath)
+	if err != nil {
+		return false
+	}
+
+	if strings.TrimSpace(string(contents)) != "yes" {
+		return false
+	}
+
+	contents, err = ioutil.ReadFile(nsNamePath)
+	if err != nil {
+		return false
+	}
+
+	// LXD and LXC set up AppArmor namespaces starting with "lxd-" and
+	// "lxc-", respectively. Return false for all other namespace
+	// identifiers.
+	name := strings.TrimSpace(string(contents))
+	if !strings.HasPrefix(name, "lxd-") && !strings.HasPrefix(name, "lxc-") {
+		return false
+	}
+	return true
+}
+
+func loadAppArmorProfiles() error {
+	candidates, err := filepath.Glob(dirs.SnapAppArmorDir + "/*")
+	if err != nil {
+		return err
+	}
+
+	profiles := make([]string, 0, len(candidates))
+	for _, profile := range candidates {
+		// Filter out profiles with names ending with ~, those are
+		// temporary files created by snapd.
+		if strings.HasSuffix(profile, "~") {
+			continue
+		}
+		profiles = append(profiles, profile)
+	}
+	if len(profiles) == 0 {
+		logger.Noticef("No profiles to load")
+		return nil
+	}
+	logger.Noticef("Loading profiles %v", profiles)
+	err = apparmor_sandbox.LoadProfiles(profiles, apparmor_sandbox.SystemCacheDir, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
@mardy do we have a helper doing this in sandbox now?

Reply:
Not right now. We have it in #11722 (and precisely here: https://github.com/snapcore/snapd/pull/11722/commits/ec08afc7d35dc9b4520e2559b8c7de8482ca34d1, search for `ReloadAllSnapProfiles`), it's under review. It does not exclude `~` files and it passes the `SkipReadCache` flag.

Reply:
Yes, skipping files ending in `~` won't harm. I'll add that to #11722 .